go/pools/smartconnpool/waitlist.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package smartconnpool
+
+import (
+	"context"
+	"sync"
+
+	"vitess.io/vitess/go/list"
+)
+
+// waiter represents a client waiting for a connection in the waitlist
+type waiter[C Connection] struct {
+	// setting is the connection Setting that we'd like, or nil if we'd like a
+	// a connection with no Setting applied
+	setting *Setting
+	// conn will be set by another client to hand over the connection to use
+	conn *Pooled[C]
+	// ctx is the context of the waiting client to check for expiration
+	ctx context.Context
+	// sema is a synchronization primitive that allows us to block until our request
+	// has been fulfilled
+	sema semaphore
+	// age is the amount of cycles this client has been on the waitlist
+	age uint32
+}
+
+type waitlist[C Connection] struct {
+	nodes sync.Pool
+	mu    sync.Mutex
+	list  list.List[waiter[C]]
+}
+
+// waitForConn blocks until a connection with the given Setting is returned by another client,
+// or until the given context expires.
+// The returned connection may _not_ have the requested Setting. This function can
+// also return a `nil` connection even if our context has expired, if the pool has
+// forced an expiration of all waiters in the waitlist.
+func (wl *waitlist[C]) waitForConn(ctx context.Context, setting *Setting) (*Pooled[C], error) {
+	elem := wl.nodes.Get().(*list.Element[waiter[C]])
+	elem.Value = waiter[C]{setting: setting, conn: nil, ctx: ctx}
+
+	wl.mu.Lock()
+	// add ourselves as a waiter at the end of the waitlist
+	wl.list.PushBackValue(elem)
+	wl.mu.Unlock()
+
+	// block on our waiter's semaphore until somebody can hand over a connection to us
+	elem.Value.sema.wait()
+
+	// we're awake -- the conn in our waiter contains the connection that was handed
+	// over to us, or nothing if we've been waken up forcefully. save the conn before
+	// we return our waiter to the pool of waiters for reuse.
+	conn := elem.Value.conn
+	wl.nodes.Put(elem)
+
+	return conn, ctx.Err()
+}
+
+// expire removes and wakes any expired waiter in the waitlist.
+// if force is true, it'll wake and remove all the waiters.
+func (wl *waitlist[C]) expire(force bool) (waiting int) {
```

Comment:
This method is only called in `pool.open` and not using `waiting` count.

Reply:
Yeah cleaned up